src/reduxTypes.js

```diff
@@ -337,6 +337,14 @@ export type GlobalState = {|
   users: UsersState,
 |};
 
+// No substate should allow `undefined`; our use of AsyncStorage
+// depends on it. (This check will also complain on `null`, which I
+// don't think we'd have a problem with. We could try to write this
+// differently if we want to allow `null`.)
+type NonMaybeProperties<O: { ... }> = $ObjMap<O, <V>(V) => $NonMaybeType<V>>;
+type NonMaybeGlobalState = NonMaybeProperties<GlobalState>;
+((s: NonMaybeGlobalState) => undefined: GlobalState => void);
```

Comment:
I'm sort of curious how this works — just defining this type, but not using it puts constraints on `GlobalState`? That seems like some spooky action-at-a-distance to me, but probably I just don't understand flow.

Reply:
Yeah, I'd be happy (and not too surprised) if we found a good, clearer alternative. The fundamental idea is that we're doing a type cast that will fail if `GlobalState` is wrong. We use that idea in `exampleData`:

```js
// Ensure every `eg.action.foo` is some well-typed action.  (We don't simply
// annotate `action` itself, because we want to keep the information of
// which one has which specific type.)
/* eslint-disable-next-line no-unused-expressions */
(action: {| [string]: Action |});
```

Since casting requires a concrete value at runtime, and we're not looking at a value of type `GlobalState`—we want to check `GlobalState` itself—I write a simple function expression, with its input annotated as desired, and cast the value of that. (The function expression is evaluated at runtime, but the function is never actually called.) This expansion of the last line might be clearer:

```js
const func1 = (s: NonMaybeGlobalState) => undefined;
type Func2 = GlobalState => void;
(func1: Func2); // This cast will fail if `GlobalState` is wrong.
```

Or, could have done

```js
const func1 = (s: NonMaybeGlobalState) => undefined;
const func2 = (s: GlobalState) => undefined;
(func1: typeof func2); // This cast will fail if `GlobalState` is wrong.
```

or

```js
type Func1 = NonMaybeGlobalState => void;
type Func2 = GlobalState => void;
const func1: Func1 = s => undefined;
(func1: Func2); // This cast will fail if `GlobalState` is wrong.
```

or even

```js
type Func1 = NonMaybeGlobalState => void;
type Func2 = GlobalState => void;
const func1: Func1 = s => undefined;
const func2: Func2 = s => undefined; // eslint-disable-line no-unused-vars
(func1: Func2); // This cast will fail if `GlobalState` is wrong.
```

Reply:
> Since casting requires a concrete value at runtime

I guess another way we could do it is to fool Flow into thinking we _do_ have a value of type `GlobalState` right in front of us:

```js
// $FlowIgnore[incompatible-type]
const s: GlobalState = 'nothing-to-see-here';
(s: NonMaybeGlobalState);
```

Reply:
The original version here was correct, but definitely a bit tricky to understand :slightly_smiling_face: 

> just defining this type, but not using it puts constraints on `GlobalState`? That seems like some spooky action-at-a-distance to me

In that original version it's important that it's not just defining a type -- as Chris said, it's defining a value (albeit a very boring function), and casting it. In general if you *just* define a type my experience has been that Flow may not bother to reason through all the implications and detect that the type can't possibly work; but once values are flowing around, it'll take notice.

One way of thinking about what that original version is doing -- and this is basically an explanation of variance, the same concept as at that article in Chris's last comment -- is:
* The function on the left demands a `NonMaybeGlobalState`.
* The cast says, let's treat it as a function that demands a `GlobalState`.
* The only way that can work is if every `GlobalState` can be used as a `NonMaybeGlobalState` -- i.e. if `GlobalState` is a subtype of `NonMaybeGlobalState`.
* So that's exactly what Flow checks when you take a function of one type and try to use it as another function type. (That plus a check on the return values.)
* Note that that subtype relationship between the parameters runs in the opposite direction from the one between the function types. We say that the function's parameters' types are *contravariant* on the function type itself.

In my review below I suggested an alternate simpler version: https://github.com/zulip/zulip-mobile/pull/4694#discussion_r654042192
```
// This function definition will fail typechecking if GlobalState is wrong.
(s: GlobalState): NonMaybeGlobalState => s; // eslint-disable-line no-unused-expressions
```